src/mechanics/mesh/MeshGmsh.cpp

```diff
@@ -0,0 +1,378 @@
+#include "MeshGmsh.h"
+
+#include "base/Exception.h"
+#include "mechanics/interpolation/InterpolationTrussLinear.h"
+#include "mechanics/interpolation/InterpolationTriangleLinear.h"
+#include "mechanics/interpolation/InterpolationQuadLinear.h"
+
+#include <array>
+#include <climits>
+#include <algorithm>
+#include <fstream>
+#include <functional>
+#include <iostream>
+#include <map>
+#include <memory>
+#include <vector>
+
+
+// Helper structs (cpp only)
+// %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+struct GmshHeader
+{
+    double version;
+    bool isBinary;
+    int doubleSize;
+};
+
+struct GmshNode
+{
+    int id = 0;
+    std::array<double, 3> coordinates;
+};
+
+struct GmshElement
+{
+    int id = 0;
+    int type = 0;
+    std::vector<int> tags;
+    std::vector<int> nodes;
+};
+
+struct GmshPhysicalNames
+{
+    int id = 0;
+    int dimension = 0;
+    std::string name;
+};
+
+
+struct GmshFileContent
+{
+    GmshHeader header;
+    int minNodeId = INT_MAX;
+    int maxNodeId = INT_MIN;
+    int dimension = 1;
+    std::vector<GmshNode> nodes;
+    std::vector<GmshElement> elements;
+    std::vector<GmshPhysicalNames> physicalNames;
+};
+
+// Helper functions (cpp only)
+// %%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
+
+static GmshHeader ReadGmshHeader(std::ifstream& file)
+{
+    std::string line;
+    getline(file, line);
+
+    int binary;
+    GmshHeader header;
+    file >> header.version;
+    file >> binary;
+    file >> header.doubleSize;
+
+    header.isBinary = (binary == 1);
+
+    std::getline(file, line); // endl
+    if (header.version < 2.)
+    {
+        throw NuTo::Exception(__PRETTY_FUNCTION__, "Gmsh version 2.0 or higher requiered. - File version is " +
+                                                           std::to_string(header.version));
+    }
+    getline(file, line); // $EndMeshFormat
+    return header;
+}
+
+
+void ReadNodesASCII(std::istream& file, GmshFileContent& fileContent)
```

Comment:
I'd prefer this to be `std::pair<std::vector<GmshNodes>, double> ReadNodesASCII(std::istream& file)`, and the call then `std::tie(fileContent.nodes, fileContent.dimension) = ReadNodesASCII(file);`. This also gives you the opportunity to include `pair` and `utility`, helping you in your quest to include all standard headers in a single file 👻 

 Similar arguments could be made for the other functions.

Reply:
Sorry to disappoint you, but pair and utility are already included by other STL headers. :trollface:  - I remove the unnecessary STL headers. We forgot that during clean up.

I also fulfill your other wish :stuck_out_tongue: